sensu/plugins/metric-os-api.py

```diff
@@ -0,0 +1,161 @@
+
+
+def request_token(url, method='GET', retries=2, **kwargs):
+    r = None
+    try:
+        for i in range(retries + 1):
+            if i > 0:
+                time.sleep(2)
+            r = requests.request(method, url, **kwargs)
+            if r.status_code < 500:
+                break
+    except requests.exceptions.RequestException as e:
+        print "{0} returned {1}".format(url, e)
+        state = STATE_WARNING
+
+    return r
+
+def request(scheme, url, timestamp, method='GET', retries=2, **kwargs):
+    global state
+    global metric_data
+
+    start = time.time()
+    r = None
+    try:
+        for i in range(retries + 1):
+            if i > 0:
+                time.sleep(2)
+            r = requests.request(method, url, **kwargs)
+            if r.status_code:
+                break
+    except requests.exceptions.RequestException as e:
+        print "{0} returned {1}".format(url, e)
+        state = STATE_WARNING
+
+    if r is None:
+        status = 0
+        status_code = 600
+        time_taken = -10
+        if state == STATE_OK:
+            state = STATE_WARNING
+    elif r.status_code != 200:
+        status = 0
+        status_code = r.status_code
+        time_taken = -5
+        state = STATE_CRITICAL
+    else:
+        status = 1
+        status_code = r.status_code
+        time_taken = round(time.time() - start, 3)
+
+
+    metric_data += '{0}.time {1} {2}\n'.format(scheme, time_taken, timestamp)
+    metric_data += '{0}.code {1} {2}\n'.format(scheme, status_code, timestamp)
+    metric_data += '{0}.status {1} {2}\n'.format(scheme, status, timestamp)
+    return r
+
+def main():
+    timestamp = int(time.time())
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument('-u', '--user', default=os.environ['OS_USERNAME'])
+    parser.add_argument('-p', '--password', default=os.environ['OS_PASSWORD'])
+    parser.add_argument('-t', '--tenant', default=os.environ['OS_TENANT_NAME'])
+    parser.add_argument('-a', '--auth-url', default=os.environ['OS_AUTH_URL'])
+    parser.add_argument('-S', '--service-type', default='nova', help='Service type must be provided')
+    parser.add_argument('-s', '--scheme', default=DEFAULT_SCHEME)
+    args = parser.parse_args()
+
+    if len(sys.argv)==1:
+        parser.print_help()
+        sys.exit(STATE_WARNING)
+
+    scheme = ''
+    if args.scheme == DEFAULT_SCHEME :
+        scheme = args.scheme + '.' + args.service_type + '.api'
+    else:
+        scheme = args.scheme + '.' + args.service_type
+
+
+    url = args.auth_url + '/tokens'
+    headers = {'Accept': 'application/json', 'Content-Type': 'application/json'}
+    data = json.dumps(
+        {
+            'auth': {
+                'tenantName': args.tenant,
+                'passwordCredentials': {
+                    'username': args.user,
+                    'password': args.password
+                }
+            }
+        })
+
+
+    r = request_token(url, 'POST', 4, data=data, headers=headers)
```

Comment:
This request a keystone token for authentication for the api request, the default for the api calls is 2, but since we want to reduce the chance of random connection issues affecting the metric collection we give it a larger number of retries than the normal api request we do later.